<a href="https://colab.research.google.com/github/mikeninerbravog/POC_ML_AI-in-Finance/blob/master/POC_ML_FraudDetection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
"""
The code includes three didactic demonstrations:

1. Fraud Detection: Application of the Isolation Forest algorithm to identify anomalous or potentially fraudulent transactions.
2. Market Trend Forecasting: Use of a Linear Regression model to estimate and project future price movements.
3. Financial Chatbot: Simulation of a virtual assistant capable of answering basic finance-related queries.
"""

import numpy as np
import pandas as pd
from sklearn.ensemble import IsolationForest
from sklearn.linear_model import LinearRegression
import random

# 1. Fraud Detection and Valid Transactions Simulation
def detect_transactions():
    np.random.seed(42)
    transactions = np.random.normal(100, 20, 100).tolist()  # Normal transactions
    transactions.extend([500, 600, 750, 1000])  # Adding cheats
    df = pd.DataFrame({'transaction_value': transactions})

    model = IsolationForest(contamination=0.05)
    df['fraudulent'] = model.fit_predict(df[['transaction_value']])

    frauds = df[df['fraudulent'] == -1]
    valid_transactions = df[df['fraudulent'] == 1]

    print("Fraudulent transactions detected:\n", frauds)
    print("Valid transactions detected:\n", valid_transactions)
    return df

# 2. Market Trend Forecast
def market_trend_prediction():
    np.random.seed(42)
    days = np.arange(1, 101)
    prices = 50 + np.cumsum(np.random.normal(0, 1, 100))  # Price simulation

    model = LinearRegression()
    model.fit(days.reshape(-1, 1), prices)
    future_prices = model.predict([[101], [102], [103]])

    print("Forecast for the next 3 days:", future_prices)
    return prices, future_prices

# 3. Simple Financial Chatbot
def chatbot_finance(question):
    """
    A simple chatbot designed to answer finance-related questions.
    This chatbot can be extended into an intelligent and autonomous assistant,
    capable of delivering more sophisticated responses and interacting dynamically with users.
    The implementation presented here serves solely as a proof of concept (PoC).
    """
    responses = {
    "What is an investment?": "An investment is when you use money to buy something with the expectation that it will be worth more in the future.",
    "How can financial fraud be avoided?": "Use secure payment methods, be cautious of offers that seem too good to be true, and monitor your transactions regularly.",
    "What is the stock market?": "It is a marketplace where people buy and sell ownership shares of companies, known as stocks."
}
    return responses.get(question, "Sorry, I didn't understand your question.")

# Performing the functions
detect_transactions()
market_trend_prediction()

# Example of interaction with the chatbot
question = "What is an investment?"
print(f"User: {question}")
print(f"Chatbot: {chatbot_finance(question)}")


Fraudulent transactions detected:
      transaction_value  fraudulent
31          137.045564          -1
74           47.605098          -1
100         500.000000          -1
101         600.000000          -1
102         750.000000          -1
103        1000.000000          -1
Valid transactions detected:
     transaction_value  fraudulent
0          109.934283           1
1           97.234714           1
2          112.953771           1
3          130.460597           1
4           95.316933           1
..                ...         ...
95          70.729701           1
96         105.922406           1
97         105.221105           1
98         100.102269           1
99          95.308257           1

[98 rows x 2 columns]
Forecast for the next 3 days: [37.12727563 36.9992055  36.87113536]
User: What is an investment?
Chatbot: An investment is when you use money to buy something with the expectation that it will be worth more in the future.




---


### **Code Breakdown**

The code presents three educational demonstrations:

1. **Fraud Detection**: Uses the *Isolation Forest* algorithm to identify suspicious transactions.
2. **Market Trend Prediction**: Implements a *Linear Regression* model to forecast future prices.
3. **Financial Chatbot**: Simulates a virtual assistant that answers basic financial questions.

---

### **Code Explanation**

#### **1. Fraud Detection and Valid Transactions**
The function `detect_transactions()` simulates financial transactions and applies **Isolation Forest**, an anomaly-based machine learning model, to classify them as **fraudulent (-1)** or **valid (1)**.

##### **How It Works**
- A dataset of **100 normal transactions** is generated using a **normal distribution** (mean = 100, standard deviation = 20).
- Four **outlier transactions** (**500, 600, 750, 1000**) are manually added to simulate fraud.
- The **Isolation Forest** model is trained with **5% contamination**, assuming around 5% of the data consists of anomalies.
- The model assigns a **fraudulent (-1)** or **valid (1)** label to each transaction.

##### **Results**
- **Fraudulent Transactions Detected**: Transactions **500, 600, 750, and 1000** were identified as fraudulent because they are significantly **higher than the average transaction value (100)**, making them outliers.
- Some additional values (**137.04 and 47.60**) were also flagged as fraudulent, possibly due to being at the extreme ends of the distribution.
- **Valid Transactions**: The rest of the transactions, which are closer to the normal range (mean = 100, std = 20), were classified as legitimate.

---

#### **2. Market Trend Prediction**
The function `market_trend_prediction()` uses **Linear Regression** to predict stock prices over the next three days.

##### **How It Works**
- A **synthetic dataset** of **100 days of stock prices** is created.
- Each price is computed using a **cumulative sum** of normally distributed random values (mean = 0, std = 1).
- The **Linear Regression** model is trained using the generated dataset.
- Predictions for **days 101, 102, and 103** are made.

##### **Results**
The model forecasts:
- **Day 101:** 37.13
- **Day 102:** 37.00
- **Day 103:** 36.87

This indicates a **slight downward trend**, based on historical price data.

---

#### **3. Simple Financial Chatbot**
The function `chatbot_finance(question)` is a **basic rule-based chatbot** that answers common financial questions.

##### **How It Works**
- A dictionary stores **predefined responses** for three financial questions:
  - **"What is an investment?"** → "Investment is when you use money to buy something expecting it to be worth more in the future."
  - **"How to avoid financial fraud?"** → "Use secure cards, be wary of deals that seem too good, and monitor transactions."
  - **"What is the stock market?"** → "It’s where people buy and sell shares of companies."
- If the user's question is not in the dictionary, it responds with: *"Sorry, I didn’t understand your question."*

##### **Example Interaction**
- **User:** *"What is an investment?"*
- **Chatbot Response:** *"Investment is when you use money to buy something expecting it to be worth more in the future."*

This chatbot serves as a **proof of concept (PoC)** and can be expanded into a more sophisticated AI-powered assistant.

---

### **Final Thoughts**
This code demonstrates three practical applications of AI in finance:
1. **Anomaly detection** for fraud prevention.
2. **Predictive analytics** for stock market forecasting.
3. **Conversational AI** for automating financial inquiries.

Each function serves as a **building block** for real-world financial applications.